src/Microsoft.ML.Data/Transforms/ValueMapping.cs

```diff
@@ -396,29 +414,32 @@ internal sealed class Options
         }
 
         internal ValueMappingTransformer(IHostEnvironment env, IDataView lookupMap,
-            string keyColumn, string valueColumn, (string outputColumnName, string inputColumnName)[] columns)
+            DataViewSchema.Column lookupKeyColumn, DataViewSchema.Column lookupValueColumn, (string outputColumnName, string inputColumnName)[] columns)
             : base(Contracts.CheckRef(env, nameof(env)).Register(nameof(ValueMappingTransformer)), columns)
         {
-            Host.CheckNonEmpty(keyColumn, nameof(keyColumn), "A key column must be specified when passing in an IDataView for the value mapping");
-            Host.CheckNonEmpty(valueColumn, nameof(valueColumn), "A value column must be specified when passing in an IDataView for the value mapping");
-            _valueMap = CreateValueMapFromDataView(lookupMap, keyColumn, valueColumn);
-            int valueColumnIdx = 0;
-            Host.Check(lookupMap.Schema.TryGetColumnIndex(valueColumn, out valueColumnIdx));
-            ValueColumnMetadata = lookupMap.Schema[valueColumnIdx].Annotations;
+            _lookupKeyColumn = lookupKeyColumn;
+            _lookupValueColumn = lookupValueColumn;
+            _valueMap = CreateValueMapFromDataView(lookupMap, _lookupKeyColumn, _lookupValueColumn);
+            ValueColumnMetadata = lookupMap.Schema[_lookupValueColumn.Index].Annotations;
 
             // Create the byte array of the original IDataView, this is used for saving out the data.
-            _dataView = GetBytesFromDataView(Host, lookupMap, keyColumn, valueColumn);
+            _dataView = GetBytesFromDataView(Host, lookupMap, lookupKeyColumn.Name, lookupValueColumn.Name);
         }
 
-        private ValueMap CreateValueMapFromDataView(IDataView dataView, string keyColumn, string valueColumn)
+        private ValueMap CreateValueMapFromDataView(IDataView dataView, DataViewSchema.Column keyColumn, DataViewSchema.Column valueColumn)
         {
-            // Confirm that the key and value columns exist in the dataView
-            Host.Check(dataView.Schema.TryGetColumnIndex(keyColumn, out int keyIdx), "Key column " + keyColumn + " does not exist in the given dataview");
-            Host.Check(dataView.Schema.TryGetColumnIndex(valueColumn, out int valueIdx), "Value column " + valueColumn + " does not exist in the given dataview");
-            var keyType = dataView.Schema[keyIdx].Type;
-            var valueType = dataView.Schema[valueIdx].Type;
-            var valueMap = ValueMap.Create(keyType, valueType, ValueColumnMetadata);
-            using (var cursor = dataView.GetRowCursor(dataView.Schema[keyIdx], dataView.Schema[valueIdx]))
+            var retrievedKeyColumn = dataView.Schema[keyColumn.Index];
```

Comment:
>Schema [](start = 46, length = 6)

Can this be out of range?

Reply:
Will add a check in the next iteration.

---
In reply to: [271535033](https://github.com/dotnet/machinelearning/pull/3180#discussion_r271535033) [](ancestors = 271535033)